ansible/roles/sunbird-system-init/templates/system_init_commands.j2

```diff
@@ -0,0 +1,336 @@
+if [ -z "$system_settings_value" ];
+then
+        echo "Unable to connect to Cassandra. Exiting program...";
+        exit 1;
+fi
+if [ "$system_settings_value" = "SYSTEM_INITIALISED" ];
+then
+        echo "System already Initialised. Exiting program...";
+        exit 0;
+fi
+
+echo "Starting System Initialisation...";
+getExistingCustodianTenantId;
+if [ -z "$sunbird_custodian_tenant_id" ];
+then
+        echo "Creating Custodian Organisation in Cassandra...";
+        getNewCustodianTenantId;
+        createOrgInCassandra;
+        getExistingCustodianTenantId;
+        if [ -z "$sunbird_custodian_tenant_id" ];
+        then
+                echo "Custodian Organisation creation FAILED in Cassandra.";
+                exit 1;
+        else
+                echo "Custodian Organisation is created in Cassandra.";
+        fi
+else
+        echo "Already a Tenant Organisation exists with the configured channel - ""$sunbird_custodian_tenant_channel";
+        echo "Making Tenant Organisation with Id - ""$sunbird_custodian_tenant_id"" as Custodian Organisation...";
+fi
+
+fetchOrgFromES;
+isOrgFoundInES=$(echo -n "$es_fetch_org_response" | jq '.found' | tr -d "\"");
+if [ "$isOrgFoundInES" = "false" ];
+then
+        echo "Syncing Custodian Organisation data to Elastic Search...";
+        createOrgInES;
+        fetchOrgFromES;
+        isOrgFoundInES=$(echo -n "$es_fetch_org_response" | jq '.found' | tr -d "\"");
+        if [ "$isOrgFoundInES" = "true" ];
+        then
+                echo "Custodian Organisation data synced to Elastic Search.";
+                setValueInSystemSettings "systemInitialisationStatus" "CUSTODIAN_ORG_CREATED";
+        else
+                echo "System Initialisation failed. Unable to sync custodian organisation data to Elastic Search.";
+                exit 1;
+        fi
+elif [ "$isOrgFoundInES" = "true" ];
+then
+        echo "Custodian Organisation data is already available in Elastic Search.";
+        getValueFromSystemSettings "systemInitialisationStatus";
+        if [ "$system_settings_value" = "SYSTEM_UNINITIALISED" ];
+        then
+            setValueInSystemSettings "systemInitialisationStatus" "CUSTODIAN_ORG_CREATED";
+        fi
+else
+        echo "System Initialisation failed. Unable to fetch data from Elastic Search - index=searchindex, type=org";
+        exit 1;
+fi
+
+getValueFromSystemSettings "systemInitialisationStatus";
+if [ "$system_settings_value" = "CUSTODIAN_ORG_CREATED" ];
+then
+        echo "Registering Custodian Organisation Channel...";
+        registerChannel;
+        if [ "$register_channel_response_code" = "200" ];
+        then
+                echo "Custodian Organisation Channel Registered Successfully.";
+                setValueInSystemSettings "systemInitialisationStatus" "CUSTODIAN_ORG_CHANNEL_REGISTERED";
+        else
+                echo "System Initialisation failed due to Custodian Organisation Channel Registration failure.";
```

Comment:
vinaya-mac:devops vinaya$ ./system_init_commands.sh 
Starting System Initialisation...
Already a Tenant Organisation exists with the configured channel - who2
Making Tenant Organisation with Id - 01258855880130560 as Custodian Organisation...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   766  100   766    0     0  13568      0 --:--:-- --:--:-- --:--:-- 13438
Custodian Organisation data is already available in Elastic Search.
Registering Custodian Organisation Channel...
System Initialisation failed due to Custodian Organisation Channel Registration failure.

Reply:
Env values used:
sunbird_custodian_tenant_name="WHO2"
sunbird_custodian_tenant_description="A health organisation"
sunbird_custodian_tenant_channel="who2"
sunbird_root_user_firstname="Jai"
sunbird_root_user_lastname="S"
sunbird_root_user_username="jai"
sunbird_root_user_password="password"
sunbird_root_user_email="vinayab@ilimi.in"
sunbird_root_user_phone="6360294273"
#ekstep_proxy_base_url="{{ekstep_proxy_base_url}}"
#ekstep_api_key="{{ekstep_api_key}}"
ekstep_api_key="eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJkMTc1MDIwNDdlODc0ODZjOTM0ZDQ1ODdlYTQ4MmM3MyJ9.7LWocwCn5rrCScFQYOne8_Op2EOo-xTCK5JCFarHKSs"
#sunbird_es_host="{{sunbird_es_host}}"
#proto="{{proto}}"
ekstep_proxy_base_url="https://dev.open-sunbird.org"
proxy_server_name="https://dev.open-sunbird.org"
#sunbird_api_auth_token="{{sunbird_api_auth_token}}"

Reply:
The script is not able to continue if channel was previously created but hash tag registration has failed.

{
    "id": "api.channel.create",
    "ver": "1.0",
    "ts": "2018-09-12T04:02:13.754Z",
    "params": {
        "resmsgid": "a1719da0-b640-11e8-b7dc-01f8480e2534",
        "msgid": null,
        "status": "failed",
        "err": "CONSTRAINT_VALIDATION_FAILED",
        "errmsg": "Object already exists with identifier: 01258855880130560"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {}
}

Reply:
For debugging better to log response code for curl commands.

Reply:
@bvinayakumar Can you clear the variable `systemInitialisationStatus` from `system_settings` table and try simulating this scenario again. We are facing this issue because the wrong status was set in `system_settings` due to the earlier bug(which is fixed now) and so, even after the fix is committed because of the old status, the script is failing.

Reply:
I removed the value from system_settings table and still the issue is reproducible in Sunbird Dev.

devuser@dev-cassandra-1:~$ cqlsh
Connected to Test Cluster at 127.0.0.1:9042.
[cqlsh 5.0.1 | Cassandra 3.9 | CQL spec 3.4.2 | Native protocol v4]
Use HELP for help.
cqlsh> use sunbird;
cqlsh:sunbird> select * from system_settings ;

 id                         | field                      | value
----------------------------+----------------------------+----------------------------------
                emailUnique |                emailUnique |                             true
             custodianOrgId |             custodianOrgId |                     060970188800
                phoneUnique |                phoneUnique |                             true
        emailUniqueAndPhone |        emailUniqueAndPhone |                            false
        custodianOrgChannel |        custodianOrgChannel |                            who21
                        003 |           channelRegStatus |                            false
 systemInitialisationStatus | systemInitialisationStatus | CUSTODIAN_ORG_HASHTAG_REGISTERED

(7 rows)
cqlsh:sunbird> delete from system_settings where id='systemInitialisationStatus';
cqlsh:sunbird> select * from system_settings ;

 id                  | field               | value
---------------------+---------------------+--------------
         emailUnique |         emailUnique |         true
      custodianOrgId |      custodianOrgId | 060970188800
         phoneUnique |         phoneUnique |         true
 emailUniqueAndPhone | emailUniqueAndPhone |        false
 custodianOrgChannel | custodianOrgChannel |        who21
                 003 |    channelRegStatus |        false

(6 rows)
cqlsh:sunbird> quit
devuser@dev-cassandra-1:~$ vi system-init.sh 
devuser@dev-cassandra-1:~$ ./system-init.sh 
Adding systemInitialisationStatus field to System Settings table...
Starting System Initialisation...
Creating Custodian Organisation in Cassandra...
Custodian Organisation is created in Cassandra.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    73  100    73    0     0   9983      0 --:--:-- --:--:-- --:--:-- 10428
Syncing Custodian Organisation data to Elastic Search...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   802  100   154  100   648   2626  11052 --:--:-- --:--:-- --:--:-- 11172
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   744  100   744    0     0  16918      0 --:--:-- --:--:-- --:--:-- 17302
Custodian Organisation data synced to Elastic Search.
Registering Custodian Organisation Channel...
curl -o /dev/null -s -w %{http_code}n                 -X POST http://10.6.0.4:8080/learning-service/api/channel/v1/create                 -H 'authorization: Bearer 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiI2NWU2MTUxNzdkODY0MGJkYWNmMWE4MWEwM2Y5MmNjYSJ9.yST4a-kA0K-r-86m0gx45IMTTZP0ujQnjFDEjv2wU0A                 -H 'cache-control: no-cache'                 -H 'content-type: application/json'                 -d '{
                        request: {
                                channel: {
                                        name:"$sunbird_custodian_tenant_channel",
                                        description:"$sunbird_custodian_tenant_description",
                                        code:"$sunbird_custodian_tenant_id"
                                }
                        }
                    }')
200
Custodian Organisation Channel Registered Successfully.
Registering Custodian Organisation HashTag for Analytics...
200
Custodian Organisation HashTag Registered Successfully.
Creating Sunbird Root User...
dev.open-sunbird.org
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   835    0   265  100   570    556   1197 --:--:-- --:--:-- --:--:--  1197
curl -X POST https://dev.open-sunbird.org/api/user/v1/create                 -H 'Cache-Control: no-cache'                 -H 'Content-Type: application/json'                 -H 'accept: application/json'                 -H 'authorization: Bearer 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJkMTc1MDIwNDdlODc0ODZjOTM0ZDQ1ODdlYTQ4MmM3MyJ9.7LWocwCn5rrCScFQYOne8_Op2EOo-xTCK5JCFarHKSs                 -d '{
                        request: {
                                firstName:"$sunbird_root_user_firstname",
                                lastName:"$sunbird_root_user_lastname",
                                userName:"$sunbird_root_user_username",
                                password:"$sunbird_root_user_password",
                                email:"$sunbird_root_user_email",
                                phone:"$sunbird_root_user_phone",
                                channel: "$sunbird_custodian_tenant_channel",
                                emailVerified: true,
                                phoneVerified: true
                        }
                    }'
{"id":"api.user.create","ver":"v1","ts":"2018-09-14 02:43:34:934+0000","params":{"resmsgid":null,"msgid":"83dad197-9a32-4105-8916-ac0c30aca2c5","err":"EMAIL_IN_USE","status":"EMAIL_IN_USE","errmsg":"Email already exists."},"responseCode":"CLIENT_ERROR","result":{}}
Sunbird root user creation failed
devuser@dev-cassandra-1:~$ cqlsh
Connected to Test Cluster at 127.0.0.1:9042.
[cqlsh 5.0.1 | Cassandra 3.9 | CQL spec 3.4.2 | Native protocol v4]
Use HELP for help.
cqlsh> use sunbird;
cqlsh:sunbird> select * from system_settings ;

 id                         | field                      | value
----------------------------+----------------------------+----------------------------------
                emailUnique |                emailUnique |                             true
             custodianOrgId |             custodianOrgId |                     020023869440
                phoneUnique |                phoneUnique |                             true
        emailUniqueAndPhone |        emailUniqueAndPhone |                            false
        custodianOrgChannel |        custodianOrgChannel |                            who22
                        003 |           channelRegStatus |                            false
 systemInitialisationStatus | systemInitialisationStatus | CUSTODIAN_ORG_HASHTAG_REGISTERED

(7 rows)
cqlsh:sunbird> quit
devuser@dev-cassandra-1:~$ ./system-init.sh 
Starting System Initialisation...
Already a Tenant Organisation exists with the configured channel - who22
Making Tenant Organisation with Id - 020023869440 as Custodian Organisation...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   744  100   744    0     0  61963      0 --:--:-- --:--:-- --:--:-- 67636
Custodian Organisation data is already available in Elastic Search.
Registering Custodian Organisation Channel...
curl -o /dev/null -s -w %{http_code}n                 -X POST http://10.6.0.4:8080/learning-service/api/channel/v1/create                 -H 'authorization: Bearer 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiI2NWU2MTUxNzdkODY0MGJkYWNmMWE4MWEwM2Y5MmNjYSJ9.yST4a-kA0K-r-86m0gx45IMTTZP0ujQnjFDEjv2wU0A                 -H 'cache-control: no-cache'                 -H 'content-type: application/json'                 -d '{
                        request: {
                                channel: {
                                        name:"$sunbird_custodian_tenant_channel",
                                        description:"$sunbird_custodian_tenant_description",
                                        code:"$sunbird_custodian_tenant_id"
                                }
                        }
                    }')
400
System Initialisation failed due to Custodian Organisation Channel Registration failure.